### PR TITLE
Continue Reading for partially-read sequences

### PR DESCRIPTION
**Pull request by jimrandomh**
Add Continue Reading as a recommendation, to suggest you finish sequences you have started.

- [x] Handle multi-sequence books/collections
- [x] Hook up Dismiss button, for getting rid of sequences you started but don't want to finish
- [x] Client-side optimistic response/state management for the dismiss button

### CODE REVIEW

**Comment by jimrandomh** [comment]
I added a num read/num total, so it says eg "Next in Mysterious Answers: 1/14 read". Which is cool. But applied to the R:A-Z collection, it said "1/1726 read", which is way too many. The set of books with a collectionId of R:A-Z doesn't seem to match the bookIds list on the collection, so there's a data integrity issue here, with data that would otherwise be hidden/ignored is getting included in the count, where it should not be. (Presumably this is mostly made up of duplicate or older-revision sequences.)

**Comment by jimrandomh** [comment]
Having a dismiss button on recommendations introduces an interesting state management problem. The effect of the button is to change what recommendations you'll get in the future (it stops recommending that you continue reading that sequence). You want it to be removed from the recommendations immediately when you click. But you *don't* want to fetch a new set of recommendations.

The way we've handled this in the past has been with redux, which is going away and we don't want to do that anymore. We're facing a similar (though not identical) problem with tracking post-views on the Apollo 2 branch.

**Comment by jimrandomh** [comment]
This one's been waiting for review for awhile.

**Comment by Discordius** [comment]
Will take a look at it today

File: packages/lesswrong/components/posts/PostsItem2.jsx [hunk 1]
```diff
@@ -177,6 +191,45 @@ const styles = (theme) => ({
...
     }
   },
+  nextUnreadIn: {
+    color: theme.palette.grey[800],
+    
+    [theme.breakpoints.up('md')]: {
+      position: "absolute",
+      left: 42,
+      top: 30,
+      zIndex: 1000,
```
Comment (Discordius) [inline packages/lesswrong/components/posts/PostsItem2.jsx]: Use the zIndex theme file for this.

File: packages/lesswrong/components/posts/PostsItem2.jsx [hunk 2]
```diff
@@ -177,6 +191,45 @@ const styles = (theme) => ({
       display: "block"
     }
   },
+  nextUnreadIn: {
+    color: theme.palette.grey[800],
+    
+    [theme.breakpoints.up('md')]: {
+      position: "absolute",
```
Comment (Discordius) [inline packages/lesswrong/components/posts/PostsItem2.jsx]: This sure feels very fragile. I expect we will want to refactor this somehow at some points.

File: packages/lesswrong/components/posts/PostsItem2.jsx [hunk 3]
```diff
@@ -177,6 +191,45 @@ const styles = (theme) => ({
...
+      width: "100%",
+      marginTop: -10,
+    },
+    
+    "& a": {
+      color: theme.palette.primary.main,
+    },
+  },
+  sequenceImage: {
+    marginTop: -12,
```
Comment (Discordius) [inline packages/lesswrong/components/posts/PostsItem2.jsx]: Also feels very fragile here. I am confused why this has a `top: 4` but a `marginTop: -12`. Shouldn't those partially cancel out?

File: packages/lesswrong/components/posts/PostsItem2.jsx [hunk 4]
```diff
@@ -177,6 +191,45 @@ const styles = (theme) => ({
...
+    marginBottom: -12,
+    top: 4,
+    position: "relative",
+    marginLeft: -60,
+    zIndex: -1,
```
Comment (Discordius) [inline packages/lesswrong/components/posts/PostsItem2.jsx]: Also add this to the zIndex theme dictionary.

File: packages/lesswrong/components/posts/PostsItem2.jsx [hunk 5]
```diff
@@ -271,17 +339,29 @@ class PostsItem2 extends PureComponent {
...
+                    height={48}
+                    width={146}
+                  />
+                </NoSSR>
+              </div>}
           </div>
+          
           {<div className={classes.actions}>
-            <PostsPageActions post={post} vertical menuClassName={classes.actionsMenu} />
+            {resumeReading && <CloseIcon onClick={() => this.props.dismissRecommendation()}/>}
```
Comment (Discordius) [inline packages/lesswrong/components/posts/PostsItem2.jsx]: Destructure from props instead of referencing props directly.

Also, I would remove the unnecessary closure and just pass `dismissRecommendations` directly to the onClick property.

File: packages/lesswrong/components/posts/PostsSingle.jsx [hunk 6]
```diff
@@ -4,7 +4,7 @@ import React from 'react';
 
 const PostsSingle = ({params, router}, context) => {
   const version = router.location && router.location.query && router.location.query.revision
-  return <Components.PostsPage documentId={params._id} version={version} />
+  return <Components.PostsPage documentId={params._id} sequenceId={null} version={version} />
```
Comment (Discordius) [inline packages/lesswrong/components/posts/PostsSingle.jsx]: Why is this necessary?

File: packages/lesswrong/components/recommendations/RecommendationsList.jsx [hunk 7]
```diff
@@ -7,10 +7,36 @@ import withUser from '../common/withUser';
 import { defaultAlgorithmSettings } from '../../lib/collections/users/recommendationSettings.js';
 
 const withRecommendations = component => {
+  // FIXME: For some unclear reason, using a ...fragment in the 'sequence' part
```
Comment (Discordius) [inline packages/lesswrong/components/recommendations/RecommendationsList.jsx]: I expect this will be relatively straightforward to fix, but seems fine for now

File: packages/lesswrong/lib/collections/collections/helpers.js [hunk 8]
```diff
@@ -0,0 +1,19 @@
+import Collections from './collection.js';
+import Books from '../books/collection.js';
+import Sequences from '../sequences/collection.js';
+
+Collections.getAllPostIDs = async (collectionID) => {
```
Comment (Discordius) [inline packages/lesswrong/lib/collections/collections/helpers.js]: This does not account for posts in Books. We should fix that, or make sure we don't have any posts in Books anymore.

File: packages/lesswrong/lib/collections/posts/custom_fields.js [hunk 9]
```diff
@@ -379,6 +379,88 @@ addFieldsDict(Posts, {
...
+        const nextPostID = await Sequences.getNextPostID(sequenceId, post._id);
+        if (nextPostID) {
+          const nextPost = await Posts.loader.load(nextPostID);
+          return accessFilterSingle(currentUser, Posts, nextPost);
+        }
+      }
+      if (post.canonicalNextPostSlug) {
+        const nextPost = await Posts.findOne({ slug: post.canonicalNextPostSlug });
+        return accessFilterSingle(currentUser, Posts, nextPost);
+      } else if(post.canonicalSequenceId) {
```
Comment (Discordius) [inline packages/lesswrong/lib/collections/posts/custom_fields.js]: I would get rid of the `else` here since the previous case ends with a return

File: packages/lesswrong/lib/collections/posts/custom_fields.js [hunk 10]
```diff
@@ -379,6 +379,88 @@ addFieldsDict(Posts, {
...
+        const nextPostID = await Sequences.getNextPostID(post.canonicalSequenceId, post._id);
+        if (nextPostID) return null;
```
Comment (Discordius) [inline packages/lesswrong/lib/collections/posts/custom_fields.js]: There is an exclamation mark missing here. This likely means this case has never been tested and I expect there are other problems with it. We should likely exercise this case at least once before merging.

File: packages/lesswrong/lib/collections/posts/custom_fields.js [hunk 11]
```diff
@@ -379,6 +379,88 @@ addFieldsDict(Posts, {
...
+  
+  // The previous post. If a sequenceId is provided, that sequence must contain
+  // this post, and this returns the post before this one in that sequence.
+  // If no sequenceId is provided, uses this post's canonical sequence.
+  prevPost: resolverOnlyField({
+    type: "Post",
+    graphQLtype: "Post",
+    viewableBy: ['guests'],
+    graphqlArguments: 'sequenceId: String',
+    resolver: async (post, { sequenceId }, { currentUser, Posts, Sequences }) => {
```
Comment (Discordius) [inline packages/lesswrong/lib/collections/posts/custom_fields.js]: Really annoying that this shares 90% of the logic with the previous case. I think it's worth factoring this out into a single function.

File: packages/lesswrong/lib/collections/sequences/helpers.js [hunk 12]
```diff
@@ -1,17 +1,78 @@
...
+  // Sort the posts retrieved back into reading order and return them
+  const postsById = keyBy(posts, post=>post._id);
+  return _.map(allPostIds, id=>postsById[id]);
+}
+
+// Given a post ID and the ID of a sequence which contains that post, return the
+// next post in the sequence, or null if it was the last post. Does not handle
+// cross-sequence boundaries. If the given post ID is not in the sequence,
+// returns null.
+Sequences.getNextPostID = async function(sequenceId, postId) {
```
Comment (Discordius) [inline packages/lesswrong/lib/collections/sequences/helpers.js]: Maybe factor out commonalities between the two functions? Not super important, but seems better for maintainability.

File: packages/lesswrong/lib/collections/users/custom_fields.js [hunk 13]
```diff
@@ -79,6 +79,33 @@ const karmaChangeSettingsType = new SimpleSchema({
   }
 })
 
+const partiallyReadSequenceItem = new SimpleSchema({
```
Comment (Discordius) [inline packages/lesswrong/lib/collections/users/custom_fields.js]: Seems pretty important to me to have a `lastRead` date or something like that, and to capture that data early, since that will enable us to remove the recommendations after some time-horizon has passed.

File: packages/lesswrong/server/partiallyReadSequences.js [hunk 14]
```diff
@@ -0,0 +1,151 @@
...
+      partiallyReadSequences: newPartiallyReadSequences
+    },
+    unset: {},
+    validate: false,
+  });
+}
+
+const EventUpdatePartialReadStatusCallback = async (event) => {
+  if (event.name === 'post-view' && event.properties.sequenceId) {
+    // Deliberately lacks an await - this runs concurrently in the background
```
Comment (Discordius) [inline packages/lesswrong/server/partiallyReadSequences.js]: But then why is this function async? To make the type signature better I would make this function sync, since it will never return a promise.

File: packages/lesswrong/server/partiallyReadSequences.js [hunk 15]
```diff
@@ -0,0 +1,151 @@
...
+    updateSequenceReadStatusForPostRead(event.userId, event.documentId, event.properties.sequenceId);
+  }
+}
+
+addCallback('lwevents.new.async', EventUpdatePartialReadStatusCallback);
+
+// Given a user and an array of post IDs, return a dictionary from
+// postID=>bool, true if the user has read the post and false otherwise.
+const postsToReadStatuses = async (user, postIDs) => {
```
Comment (Discordius) [inline packages/lesswrong/server/partiallyReadSequences.js]: I will trust you that this works since this kind of code is a pain to review.

File: packages/lesswrong/server/recommendations.js [hunk 16]
```diff
@@ -159,18 +160,78 @@ const getRecommendations = async ({count, algorithm, currentUser}) => {
...
+const getResumeSequences = async (currentUser, context) => {
+  if (!currentUser)
+    return [];
+  if (!currentUser.partiallyReadSequences)
+    return [];
+  
+  return Promise.all(_.map(currentUser.partiallyReadSequences,
+    async partiallyReadSequence => ({
+      sequence: partiallyReadSequence.sequenceId
+        ? await context["Sequences"].loader.load(partiallyReadSequence.sequenceId)
```
Comment (Discordius) [inline packages/lesswrong/server/recommendations.js]: I think I prefer this to a proper function instead of an object construction, but it's fine this way. Just seemed a bit hard to read.

File: packages/lesswrong/server/recommendations.js [hunk 17]
```diff
@@ -159,18 +160,78 @@ const getRecommendations = async ({count, algorithm, currentUser}) => {
...
+        : null,
+      collection: partiallyReadSequence.collectionId
+        ? await context["Collections"].loader.load(partiallyReadSequence.collectionId)
```
Comment (Discordius) [inline packages/lesswrong/server/recommendations.js]: I would deconstruct these in the function signature instead of referencing them this way.

File: packages/lesswrong/server/recommendations.js (hunk 16 above)
Comment (Discordius) [inline packages/lesswrong/server/recommendations.js]: I would probably deconstruct `partiallyReadSequence` before you go into this case. I expect this to make this a lot easier to understand.

**Comment by Discordius** [comment]
Left a lot of comments, some of them minor, some of them major. I think you can address a lot of them, but some of them might be better addressed with us going over the code together tomorrow. 

Overall architecture seems fine I think.

**Comment by Discordius** [comment]
After testing the UI one more problem: 

The username breaks out of bounds during SSR. Looks like this: 

Screen Shot 2019-06-04 at 3 51 29 PM

**Comment by Discordius** [comment]
My guess is that we can likely avoid the `NoSSR` thing and display the image directly by constructing the image URL directly, since in this case we don't really need any fancy responsiveness.

**Comment by Discordius** [comment]
In terms of UI improvements, I would also change the font of the "Next Unread" text to sans-serif, since I think our serif is slightly too hard to read at that font-size.

**Comment by Discordius** [comment]
On mobile, there are some padding and margin problems, I think caused by the image being talled than the usual second line: 

Screen Shot 2019-06-04 at 3 53 59 PM

**Comment by Discordius** [comment]
In terms of user experience, it feels particularly important that the continue reading updates when you revisit the frontpage (much more than the recommendations), because I presumably often do that after I have made some progress in the sequence, and the old recommendation still being there would probably make me think the recommendation system is broken.

**Comment by Discordius** [comment]
Other UI things: 

+ The dismiss button really needs a Tooltip, since I had no idea what it meant
+ The dismiss button is not accessible at all on mobile, which seems bad

File: packages/lesswrong/components/posts/PostsItemTitle.jsx [hunk 18]
```diff
@@ -10,7 +10,7 @@ const styles = theme => ({
   root: {
     color: "rgba(0,0,0,.87)",
     position: "relative",
-    lineHeight: "2rem",
+    lineHeight: "1.5rem",
```
Comment (Discordius) [inline packages/lesswrong/components/posts/PostsItemTitle.jsx]: Changing this line-height broke the vertical-centering of the title text, causing all posts to be vertically offset from the center. We should likely ensure vertical centering with flexbox or the vertical-alignment property.

File: packages/lesswrong/components/posts/PostsSingle.jsx (hunk 6 above)
Comment (jimrandomh) [reply packages/lesswrong/components/posts/PostsSingle.jsx]: `sequenceId` is context used to determine which sequence's navigation to show on the top/bottom (it needs to be specified separately from the post, since the same post can be in more than one sequence). This component corresponds to a route where no sequence was specified, hence null (the resolver receives the null and fills in the canonical sequence if appropriate). `PostsPage` uses `extraVariables` with the `withDocument` HoC, which (empirically) requires all `extraVariables` to be present; omitting this parameter wouldn't make an implicit null, it would just crash.

File: packages/lesswrong/components/posts/PostsItem2.jsx (hunk 3 above)
Comment (jimrandomh) [reply packages/lesswrong/components/posts/PostsItem2.jsx]: The `marginTop` cancels out the top-padding of `.PostsItem2-postsItem`. Top `top: 4` is empirically necessary, but I don't have a principled reason why. But... it looks like specifying the height on both the image and its container (rather than just the image, makes it line up correctly again; there were 4px of height mismatch that were neither margin, padding, nor border separating the image's height from the container's height.

File: packages/lesswrong/components/posts/PostsItemTitle.jsx (hunk 18 above)
Comment (jimrandomh) [reply packages/lesswrong/components/posts/PostsItemTitle.jsx]: If you mean the vertical offset in the continue-reading entries, that's deliberate (and it comes from a `top: -5` line in `hasResumeReading`/`& $title`, not from here). The offset is necessary to make room for the second line below. If you're referring to something else, I can't find it; everything looks correctly aligned to me, and changing the value of this `line-height` in the inspector doesn't cause anything to move.

File: packages/lesswrong/components/posts/PostsItem2.jsx (hunk 5 above)
Comment (jimrandomh) [reply packages/lesswrong/components/posts/PostsItem2.jsx]: It isn't necessary in this case, but as convention, I prefer we always make a closure when using an onFoo attribute, because it's way too easy to end up with `this` bound incorrectly otherwise..

File: packages/lesswrong/lib/collections/posts/custom_fields.js (hunk 11 above)
Comment (jimrandomh) [reply packages/lesswrong/lib/collections/posts/custom_fields.js]: I actually don't think so. It's symmetrical, but, in order to refactor, the separated function would be taking field-names as part of its signature, which is terrible.

File: packages/lesswrong/lib/collections/sequences/helpers.js (hunk 12 above)
Comment (jimrandomh) [reply packages/lesswrong/lib/collections/sequences/helpers.js]: By line count, most of this is bounds checks, which are different. It's possible to factor out a common function, but it would have to use a `delta` parameter, which is a more confusing pattern than this particular function pair is worth.

File: packages/lesswrong/lib/collections/users/custom_fields.js (hunk 13 above)
Comment (jimrandomh) [reply packages/lesswrong/lib/collections/users/custom_fields.js]: We have that data, just not denormalized here--it's just the read-date of `lastReadPostId`, which is in `lwevents` and `readstatuses`.

File: packages/lesswrong/components/posts/PostsItem2.jsx (hunk 2 above)
Comment (jimrandomh) [reply packages/lesswrong/components/posts/PostsItem2.jsx]: To the extent this is fragile, I think the fragility is in the conceptual layout, not anything the code is adding.

File: packages/lesswrong/lib/collections/collections/helpers.js (hunk 8 above)
Comment (jimrandomh) [reply packages/lesswrong/lib/collections/collections/helpers.js]: Yep, this is a bug. Will address before merging.

**Comment by jimrandomh** [comment]
[Editing your top-level comments to have checkboxes, so I can keep track of and indicate which are addressed]

File: packages/lesswrong/components/posts/PostsItem2.jsx (hunk 5 above)
Comment (Discordius) [reply packages/lesswrong/components/posts/PostsItem2.jsx]: Hmm, ok. That seems fine.

**Comment by jimrandomh** [comment]
Some nice-to-have polish that was requested in this PR, but which wasn't important enough to block merge (particularly given that this is admin-only), is now in #2024.

**Comment by quinn-dougherty** [comment]
(Instead of possibly polluting issues, allow me to necro this...)

I don't use this functionality cuz the binary checkmark on post I've simply opened and may have skimmed some sentences of is very annoying. 

I think substack did a great job with percentages.
